internal/api/impl/auth/service.go

```diff
@@ -0,0 +1,119 @@
+// Copyright 2024 The Perses Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package auth
+
+import (
+	"errors"
+
+	"github.com/perses/perses/internal/api/interface/v1/user"
+	databaseModel "github.com/perses/perses/internal/api/shared/database/model"
+	v1 "github.com/perses/perses/pkg/model/api/v1"
+)
+
+// arbitrate decides if we take the old or new string.
+// Return a boolean saying if the result is different from old value.
+func arbitrate(old, new string) (string, bool) {
+	if len(new) > 0 {
+		return new, old != new
+	}
+	return old, false
+}
+
+// saveProfileInfo build a user spec merging the old spec with a given user profile
+// Return a boolean saying if the result is different from old value.
+func saveProfileInfo(old v1.UserSpec, uInfoProfile externalUserInfoProfile) (v1.UserSpec, bool) {
+	hasChanged := false
+	singleChange := false
+	old.FirstName, singleChange = arbitrate(old.FirstName, uInfoProfile.GivenName)
+	hasChanged = hasChanged || singleChange
+	old.LastName, singleChange = arbitrate(old.LastName, uInfoProfile.FamilyName)
+	hasChanged = hasChanged || singleChange
+	return old, hasChanged
+}
+
+// saveProviderInfo takes provider context of external user info and control that there is not already a different
+// provider registered for that user.
+// Return a boolean saying if the result is different from old value.
+// If it is a case then an error is returned.
+// NB: Currently done like this for sake of simplicity. This is the most straightforward way to ensure that we can't
+// impersonate another user.
+func saveProviderInfo(old v1.UserSpec, uInfoProvider v1.OAuthProvider) (v1.UserSpec, bool, error) {
+	if len(old.NativeProvider.Password) > 0 {
+		return old, false, errors.New("this user is already registered with the native provider")
+	}
+	foundPerfectMatch := false
+	for _, provider := range old.OauthProviders {
+		if provider != uInfoProvider {
+			return old, false, errors.New("this user is already registered with a different oauth provider context")
```

Comment:
I think it will be nice if this error is not lost because it will be hide with the `InternalServerError`

Reply:
perhaps in that particular case we could have an info or debug log ?

Reply:
we have already an error log on top when the sync does not work. 

![image](https://github.com/perses/perses/assets/10258608/9cf0ae3d-ed1b-42ba-ac20-27739410a432)

In this case, I was already logged in with celian.garcia1@gmail.com on google so I was not able to use the same email for gitlab.